File: src/remedy/connect/store.py
"""Paired-device store and pause flag under ``~/.remedy/auth/connect``.

Device records are DPAPI-sealed on Windows and mode ``0600`` elsewhere.
Honors ``REMEDY_HOME``.
"""

from __future__ import annotations

import base64
import contextlib
import json
import logging
import re
import threading
import time
from pathlib import Path
from typing import Any

from remedy.core.atomic_json import write_json_atomic
from remedy.home import default_home

logger = logging.getLogger(__name__)

DEVICE_ID_RE = re.compile(r"^[a-f0-9]{16,64}$")
MAX_DEVICES = 3

_lock = threading.RLock()


def connect_root(home: Path | str | None = None) -> Path:
    base = Path(home) if home is not None else default_home()
    d = base / "auth" / "connect"
    d.mkdir(parents=True, exist_ok=True)
    _harden(d, is_dir=True)
    return d


def devices_dir(home: Path | str | None = None) -> Path:
    d = connect_root(home) / "devices"
    d.mkdir(parents=True, exist_ok=True)
    _harden(d, is_dir=True)
    return d


def state_path(home: Path | str | None = None) -> Path:
    return connect_root(home) / "state.json"


def _harden(path: Path, *, is_dir: bool) -> None:
    with contextlib.suppress(OSError):
        path.chmod(0o700 if is_dir else 0o600)
    try:
        from remedy.interfaces.secret_store import _harden_path

        _harden_path(path, is_dir=is_dir)
    except Exception:
        pass


def _write_sealed_json(path: Path, payload: dict[str, Any]) -> None:
    plain = json.dumps(payload, indent=2, sort_keys=True).encode("utf-8")
    envelope: dict[str, Any] = {
        "v": 2,
        "encoding": "plain",
        "payload": payload,
    }
    try:
        from remedy.interfaces.secret_store import _dpapi_available, _dpapi_protect

        if _dpapi_available():
            sealed = _dpapi_protect(plain)
            envelope = {
                "v": 2,
                "encoding": "dpapi",
                "dpapi": base64.b64encode(sealed).decode("ascii"),
            }
    except Exception as exc:
        logger.warning("connect DPAPI protect failed; storing owner-only plain: %s", exc)
    write_json_atomic(path, envelope, mode=0o600)
    _harden(path, is_dir=False)


def _read_sealed_json(path: Path) -> dict[str, Any] | None:
    try:
        raw = path.read_bytes()
    except OSError:
        return None
    if not raw:
        return None
    try:
        outer = json.loads(raw.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError):
        return None
    if not isinstance(outer, dict):
        return None
    encoding = str(outer.get("encoding") or "").strip().lower()
    if encoding == "dpapi" or outer.get("dpapi"):
        blob = outer.get("dpapi") or outer.get("payload") or ""
        try:
            from remedy.interfaces.secret_store import _dpapi_unprotect

            plain = _dpapi_unprotect(base64.b64decode(str(blob)))
            inner = json.loads(plain.decode("utf-8"))
        except Exception as exc:
            logger.warning("connect DPAPI unprotect failed: %s", exc)
            return None
        return inner if isinstance(inner, dict) else None
    payload = outer.get("payload")
    if isinstance(payload, dict):
        return payload
    # Legacy / unwrapped device document.
    if "id" in outer or "public_hex" in outer:
        return outer
    return None


def _device_path(device_id: str, home: Path | str | None = None) -> Path:
    if not DEVICE_ID_RE.fullmatch(device_id):
        raise ValueError("invalid device id")
    return devices_dir(home) / f"{device_id}.json"


def save_device(record: dict[str, Any], home: Path | str | None = None) -> dict[str, Any]:
    device_id = str(record.get("id") or "").strip().lower()
    if not DEVICE_ID_RE.fullmatch(device_id):
        raise ValueError("invalid device id")
    clean = {
        "id": device_id,
        "name": str(record.get("name") or "phone")[:80],
        "public_hex": str(record.get("public_hex") or "").strip().lower(),
        "paired_at": float(record.get("paired_at") or time.time()),
        "revoked": bool(record.get("revoked", False)),
    }
    with _lock:
        _write_sealed_json(_device_path(device_id, home), clean)
    if clean["revoked"]:
        _revoked_live.add(device_id)
    else:
        _revoked_live.discard(device_id)
    return clean


# Device ids revoked in this process. The gateway checks this before every
# record so a live session ends at the next frame, without a disk read.
_revoked_live: set[str] = set()


def is_revoked_live(device_id: str) -> bool:
    return str(device_id or "").strip().lower() in _revoked_live


def get_device(device_id: str, home: Path | str | None = None) -> dict[str, Any] | None:
    try:
        path = _device_path(str(device_id).strip().lower(), home)
    except ValueError:
        return None
    with _lock:
        if not path.is_file():
            return None
        return _read_sealed_json(path)


def list_devices(home: Path | str | None = None, *, include_revoked: bool = True) -> list[dict[str, Any]]:
    out: list[dict[str, Any]] = []
    with _lock:
        root = devices_dir(home)
        for path in sorted(root.glob("*.json")):
            rec = _read_sealed_json(path)
            if not rec:
                continue
            if not include_revoked and rec.get("revoked"):
                continue
            out.append(rec)
    return out


def active_device_count(home: Path | str | None = None) -> int:
    return len(list_devices(home, include_revoked=False))


def device_public_meta(home: Path | str | None = None) -> list[dict[str, Any]]:
    """Owner-visible metadata: no public keys, no secrets.

    Revoked devices are hidden so a revoke visibly removes the phone from
    Connect settings. The record stays on disk so reconnects are refused.
    """
    rows = []
    for rec in list_devices(home, include_revoked=False):
        rows.append(
            {
                "id": rec.get("id"),
                "name": rec.get("name"),
                "paired_at": rec.get("paired_at"),
                "revoked": bool(rec.get("revoked")),
            }
        )
    return rows


def revoke_device(device_id: str, home: Path | str | None = None) -> dict[str, Any] | None:
    rec = get_device(device_id, home)
    if rec is None:
        return None
    rec["revoked"] = True
    save_device(rec, home)
    _revoked_live.add(str(rec["id"]))
    try:
        from remedy.connect.audit import append_event

        append_event("revoke", home=Path(home) if home else None, device_id=rec["id"])
    except Exception:
        pass
    return rec


def find_device_by_public(public_hex: str, home: Path | str | None = None) -> dict[str, Any] | None:
    want = str(public_hex or "").strip().lower()
    if not want:
        return None
    for rec in list_devices(home, include_revoked=True):
        if str(rec.get("public_hex") or "").strip().lower() == want:
            return rec
    return None


def load_state(home: Path | str | None = None) -> dict[str, Any]:
    path = state_path(home)
    try:
        raw = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return {"paused": False}
    if not isinstance(raw, dict):
        return {"paused": False}
    return {"paused": bool(raw.get("paused", False))}


def save_state(state: dict[str, Any], home: Path | str | None = None) -> dict[str, Any]:
    clean = {"paused": bool(state.get("paused", False))}
    path = state_path(home)
    with _lock:
        write_json_atomic(path, clean, mode=0o600)
        _harden(path, is_dir=False)
    return clean


# ``is_paused`` runs before every inbound record on every phone socket; a
# short cache keeps that off the disk, and a torn read (Windows can raise
# PermissionError during an atomic replace) falls back to the last good value
# instead of momentarily reporting "not paused".
_PAUSED_TTL_S = 0.5
_paused_cache: dict[str, tuple[bool, float]] = {}


def is_paused(home: Path | str | None = None) -> bool:
    key = str(state_path(home))
    now = time.monotonic()
    hit = _paused_cache.get(key)
    if hit is not None and now - hit[1] < _PAUSED_TTL_S:
        return hit[0]
    try:
        val = bool(load_state(home).get("paused"))
    except Exception:
        val = hit[0] if hit is not None else False
    _paused_cache[key] = (val, now)
    return val


def set_paused(paused: bool, home: Path | str | None = None) -> None:
    was = is_paused(home)
    save_state({"paused": bool(paused)}, home)
    _paused_cache[str(state_path(home))] = (bool(paused), time.monotonic())
    if bool(paused) and not was:
        try:
            from remedy.connect.audit import append_event

            append_event("pause", home=Path(home) if home else None, on="1")
        except Exception:
            pass
    elif not paused and was:
        try:
            from remedy.connect.audit import append_event

            append_event("pause", home=Path(home) if home else None, on="0")
        except Exception:
            pass

File: native/go/connect/store.go
package connect

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"sort"
	"strings"
	"sync"
	"time"

	"github.com/AhmiDarrow/RemedyAI/native/go/secret"
)

const (
	// MaxDevices is the active paired-device cap (Python store.MAX_DEVICES).
	MaxDevices = 3
)

var (
	deviceIDRe  = regexp.MustCompile(`^[a-f0-9]{16,64}$`)
	storeMu     sync.Mutex
	revokedLive = map[string]struct{}{}
)

// Device is a paired-phone record under auth/connect/devices/.
type Device struct {
	ID        string  `json:"id"`
	Name      string  `json:"name"`
	PublicHex string  `json:"public_hex"`
	PairedAt  float64 `json:"paired_at"`
	Revoked   bool    `json:"revoked"`
}

// DevicesDir is ~/.remedy/auth/connect/devices.
func DevicesDir(home string) (string, error) {
	root, err := ConnectRoot(home)
	if err != nil {
		return "", err
	}
	dir := filepath.Join(root, "devices")
	if err := os.MkdirAll(dir, 0o700); err != nil {
		return "", err
	}
	_ = os.Chmod(dir, 0o700)
	return dir, nil
}

func devicePath(deviceID, home string) (string, error) {
	id := strings.ToLower(strings.TrimSpace(deviceID))
	if !deviceIDRe.MatchString(id) {
		return "", fmt.Errorf("invalid device id")
	}
	dir, err := DevicesDir(home)
	if err != nil {
		return "", err
	}
	return filepath.Join(dir, id+".json"), nil
}

// SaveDevice persists a sealed device record.
func SaveDevice(rec Device, home string) (Device, error) {
	id := strings.ToLower(strings.TrimSpace(rec.ID))
	if !deviceIDRe.MatchString(id) {
		return Device{}, fmt.Errorf("invalid device id")
	}
	name := strings.TrimSpace(rec.Name)
	if name == "" {
		name = "phone"
	}
	if len(name) > 80 {
		name = name[:80]
	}
	pairedAt := rec.PairedAt
	if pairedAt == 0 {
		pairedAt = float64(time.Now().UnixNano()) / 1e9
	}
	clean := Device{
		ID:        id,
		Name:      name,
		PublicHex: strings.ToLower(strings.TrimSpace(rec.PublicHex)),
		PairedAt:  pairedAt,
		Revoked:   rec.Revoked,
	}
	path, err := devicePath(id, home)
	if err != nil {
		return Device{}, err
	}
	storeMu.Lock()
	defer storeMu.Unlock()
	if err := writeSealedJSON(path, clean); err != nil {
		return Device{}, err
	}
	if clean.Revoked {
		revokedLive[id] = struct{}{}
	} else {
		delete(revokedLive, id)
	}
	return clean, nil
}

// GetDevice loads one device record, or nil when missing/unreadable.
func GetDevice(deviceID, home string) (*Device, error) {
	path, err := devicePath(deviceID, home)
	if err != nil {
		return nil, nil
	}
	storeMu.Lock()
	defer storeMu.Unlock()
	rec, err := readSealedDevice(path)
	if err != nil || rec == nil {
		return nil, nil
	}
	return rec, nil
}

// ListDevices returns sealed device records, optionally including revoked.
func ListDevices(home string, includeRevoked bool) ([]Device, error) {
	dir, err := DevicesDir(home)
	if err != nil {
		return nil, err
	}
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil, err
	}
	names := make([]string, 0, len(entries))
	for _, e := range entries {
		if e.IsDir() || !strings.HasSuffix(e.Name(), ".json") {
			continue
		}
		names = append(names, e.Name())
	}
	sort.Strings(names)

	storeMu.Lock()
	defer storeMu.Unlock()
	out := make([]Device, 0, len(names))
	for _, name := range names {
		rec, err := readSealedDevice(filepath.Join(dir, name))
		if err != nil || rec == nil {
			continue
		}
		if !includeRevoked && rec.Revoked {
			continue
		}
		out = append(out, *rec)
	}
	return out, nil
}

// ActiveDeviceCount is the number of non-revoked paired devices.
func ActiveDeviceCount(home string) (int, error) {
	list, err := ListDevices(home, false)
	if err != nil {
		return 0, err
	}
	return len(list), nil
}

// FindDeviceByPublic locates a device by its X25519 public key hex.
func FindDeviceByPublic(publicHex, home string) (*Device, error) {
	want := strings.ToLower(strings.TrimSpace(publicHex))
	if want == "" {
		return nil, nil
	}
	list, err := ListDevices(home, true)
	if err != nil {
		return nil, err
	}
	for i := range list {
		if strings.ToLower(strings.TrimSpace(list[i].PublicHex)) == want {
			rec := list[i]
			return &rec, nil
		}
	}
	return nil, nil
}

// RevokeDevice marks a device revoked and tracks it live.
func RevokeDevice(deviceID, home string) (*Device, error) {
	rec, err := GetDevice(deviceID, home)
	if err != nil || rec == nil {
		return nil, err
	}
	rec.Revoked = true
	saved, err := SaveDevice(*rec, home)
	if err != nil {
		return nil, err
	}
	_ = AppendAudit("revoke", home, map[string]string{"device_id": saved.ID})
	return &saved, nil
}

// DevicePublicMeta is owner-visible device metadata (no public keys / secrets).
type DevicePublicMeta struct {
	ID       string  `json:"id"`
	Name     string  `json:"name"`
	PairedAt float64 `json:"paired_at"`
	Revoked  bool    `json:"revoked"`
}

// DevicePublicMetaList returns active paired devices without public_hex.
// Revoked devices are hidden so revoke removes them from Connect settings.
func DevicePublicMetaList(home string) []DevicePublicMeta {
	list, err := ListDevices(home, false)
	if err != nil || len(list) == 0 {
		return []DevicePublicMeta{}
	}
	out := make([]DevicePublicMeta, 0, len(list))
	for _, rec := range list {
		out = append(out, DevicePublicMeta{
			ID:       rec.ID,
			Name:     rec.Name,
			PairedAt: rec.PairedAt,
			Revoked:  rec.Revoked,
		})
	}
	return out
}

// IsRevokedLive reports whether deviceID was revoked in this process.
func IsRevokedLive(deviceID string) bool {
	id := strings.ToLower(strings.TrimSpace(deviceID))
	storeMu.Lock()
	defer storeMu.Unlock()
	_, ok := revokedLive[id]
	return ok
}

const pausedCacheTTL = 500 * time.Millisecond

var (
	pausedCacheMu sync.Mutex
	pausedCache   = map[string]pausedHit{}
)

type pausedHit struct {
	val bool
	at  time.Time
}

// StatePath is ~/.remedy/auth/connect/state.json.
func StatePath(home string) (string, error) {
	root, err := ConnectRoot(home)
	if err != nil {
		return "", err
	}
	return filepath.Join(root, "state.json"), nil
}

// IsPaused reports the Connect pause flag (short-cached; torn reads keep last good).
func IsPaused(home string) bool {
	path, err := StatePath(home)
	if err != nil {
		return false
	}
	key := path
	now := time.Now()
	pausedCacheMu.Lock()
	hit, ok := pausedCache[key]
	if ok && now.Sub(hit.at) < pausedCacheTTL {
		pausedCacheMu.Unlock()
		return hit.val
	}
	pausedCacheMu.Unlock()

	val := false
	raw, readErr := os.ReadFile(path)
	if readErr == nil && len(raw) > 0 {
		var outer map[string]any
		if json.Unmarshal(raw, &outer) == nil {
			val = pausedFromEnvelope(outer)
		}
	} else if readErr != nil && !os.IsNotExist(readErr) && ok {
		val = hit.val
	}

	pausedCacheMu.Lock()
	pausedCache[key] = pausedHit{val: val, at: now}
	pausedCacheMu.Unlock()
	return val
}

func pausedFromEnvelope(outer map[string]any) bool {
	encoding, _ := outer["encoding"].(string)
	encoding = strings.ToLower(strings.TrimSpace(encoding))
	var inner map[string]any
	switch {
	case encoding == "dpapi":
		blob, _ := outer["dpapi"].(string)
		if blob == "" {
			blob, _ = outer["payload"].(string)
		}
		cipher, err := base64.StdEncoding.DecodeString(blob)
		if err != nil {
			return false
		}
		plain, err := secret.Unprotect(cipher)
		if err != nil {
			return false
		}
		if json.Unmarshal(plain, &inner) != nil {
			return false
		}
	default:
		if payload, ok := outer["payload"].(map[string]any); ok {
			inner = payload
		} else if _, has := outer["paused"]; has {
			inner = outer
		} else {
			return false
		}
	}
	paused, _ := inner["paused"].(bool)
	return paused
}

// SetPaused writes the Connect pause flag and audits transitions.
func SetPaused(paused bool, home string) error {
	was := IsPaused(home)
	path, err := StatePath(home)
	if err != nil {
		return err
	}
	payload := map[string]any{"paused": paused}
	plain, err := json.MarshalIndent(payload, "", "  ")
	if err != nil {
		return err
	}
	envelope := map[string]any{
		"v":        2,
		"encoding": "plain",
		"payload":  payload,
	}
	if runtime.GOOS == "windows" {
		if sealed, err := secret.Protect(plain); err == nil {
			envelope = map[string]any{
				"v":        2,
				"encoding": "dpapi",
				"dpapi":    base64.StdEncoding.EncodeToString(sealed),
			}
		}
	}
	raw, err := json.MarshalIndent(envelope, "", "  ")
	if err != nil {
		return err
	}
	raw = append(raw, '\n')
	storeMu.Lock()
	err = writeBytesAtomic(path, raw)
	storeMu.Unlock()
	if err != nil {
		return err
	}

	pausedCacheMu.Lock()
	pausedCache[path] = pausedHit{val: paused, at: time.Now()}
	pausedCacheMu.Unlock()

	if paused && !was {
		_ = AppendAudit("pause", home, map[string]string{"on": "1"})
	} else if !paused && was {
		_ = AppendAudit("pause", home, map[string]string{"on": "0"})
	}
	return nil
}

func writeSealedJSON(path string, payload Device) error {
	plain, err := json.MarshalIndent(payload, "", "  ")
	if err != nil {
		return err
	}
	var asMap map[string]any
	if err := json.Unmarshal(plain, &asMap); err != nil {
		return err
	}
	plain, err = json.MarshalIndent(asMap, "", "  ")
	if err != nil {
		return err
	}

	envelope := map[string]any{
		"v":        2,
		"encoding": "plain",
		"payload":  asMap,
	}
	if runtime.GOOS == "windows" {
		if sealed, err := secret.Protect(plain); err == nil {
			envelope = map[string]any{
				"v":        2,
				"encoding": "dpapi",
				"dpapi":    base64.StdEncoding.EncodeToString(sealed),
			}
		}
	}
	raw, err := json.MarshalIndent(envelope, "", "  ")
	if err != nil {
		return err
	}
	raw = append(raw, '\n')
	return writeBytesAtomic(path, raw)
}

func readSealedDevice(path string) (*Device, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	if len(raw) == 0 {
		return nil, nil
	}
	var outer map[string]any
	if err := json.Unmarshal(raw, &outer); err != nil {
		return nil, nil
	}
	encoding, _ := outer["encoding"].(string)
	encoding = strings.ToLower(strings.TrimSpace(encoding))
	var inner map[string]any
	switch {
	case encoding == "dpapi":
		blob, _ := outer["dpapi"].(string)
		if blob == "" {
			blob, _ = outer["payload"].(string)
		}
		cipher, err := base64.StdEncoding.DecodeString(blob)
		if err != nil {
			return nil, nil
		}
		plain, err := secret.Unprotect(cipher)
		if err != nil {
			return nil, nil
		}
		if err := json.Unmarshal(plain, &inner); err != nil {
			return nil, nil
		}
	default:
		if payload, ok := outer["payload"].(map[string]any); ok {
			inner = payload
		} else if _, hasID := outer["id"]; hasID {
			inner = outer
		} else if _, hasPub := outer["public_hex"]; hasPub {
			inner = outer
		} else {
			return nil, nil
		}
	}
	rec, err := deviceFromMap(inner)
	if err != nil {
		return nil, nil
	}
	return &rec, nil
}

func deviceFromMap(m map[string]any) (Device, error) {
	id, _ := m["id"].(string)
	name, _ := m["name"].(string)
	pub, _ := m["public_hex"].(string)
	var pairedAt float64
	switch v := m["paired_at"].(type) {
	case float64:
		pairedAt = v
	case json.Number:
		f, _ := v.Float64()
		pairedAt = f
	}
	revoked, _ := m["revoked"].(bool)
	if strings.TrimSpace(id) == "" {
		return Device{}, fmt.Errorf("missing id")
	}
	return Device{
		ID:        strings.ToLower(strings.TrimSpace(id)),
		Name:      name,
		PublicHex: strings.ToLower(strings.TrimSpace(pub)),
		PairedAt:  pairedAt,
		Revoked:   revoked,
	}, nil
}
